### ISSUE TITLE
[Bug] QMIX remote worker does not have GPU `(torch.cuda.is_available() == False)` but local mode works fine

### ISSUE THREAD

**Issue by mickelliu** [ray-project/ray#21217]
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

RLlib

### What happened + What you expected to happen

I got an error stating the CUDA device is unavailable while calling `make_env` on the remote worker (I think). My Gym environment requires an instantiation of a UnrealEngine binary and then loading a pre-trained CV model to process the raw observation from interacting with the binary. I need to use `torch.load()` to load this pre-trained CV model and I got a torch serialization error while loading this model.

The error report is as follows:
```
(QMIX pid=108284) 2021-12-21 17:27:52,536       ERROR worker.py:431 -- Exception raised in creation task: The actor died because of an error raised in its creation task, ray::QMIX.__init__() (pid=108284, ip=162.105.162.24)
(QMIX pid=108284)   File "/home/xjp/Miniconda3/envs/active-pose/lib/python3.8/site-packages/ray/rllib/agents/trainer_template.py", line 102, in __init__
(QMIX pid=108284)     Trainer.__init__(self, config, env, logger_creator,
(QMIX pid=108284)   File "/home/xjp/Miniconda3/envs/active-pose/lib/python3.8/site-packages/ray/rllib/agents/trainer.py", line 661, in __init__
(QMIX pid=108284)     super().__init__(config, logger_creator, remote_checkpoint_dir,
(QMIX pid=108284)   File "/home/xjp/Miniconda3/envs/active-pose/lib/python3.8/site-packages/ray/tune/trainable.py", line 121, in __init__
(QMIX pid=108284)     self.setup(copy.deepcopy(self.config))
(QMIX pid=108284)   File "/home/xjp/Miniconda3/envs/active-pose/lib/python3.8/site-packages/ray/rllib/agents/trainer_template.py", line 113, in setup
(QMIX pid=108284)     super().setup(config)
(QMIX pid=108284)   File "/home/xjp/Miniconda3/envs/active-pose/lib/python3.8/site-packages/ray/rllib/agents/trainer.py", line 764, in setup
(QMIX pid=108284)     self._init(self.config, self.env_creator)
(QMIX pid=108284)   File "/home/xjp/Miniconda3/envs/active-pose/lib/python3.8/site-packages/ray/rllib/agents/trainer_template.py", line 136, in _init
(QMIX pid=108284)     self.workers = self._make_workers(
(QMIX pid=108284)   File "/home/xjp/Miniconda3/envs/active-pose/lib/python3.8/site-packages/ray/rllib/agents/trainer.py", line 1727, in _make_workers
(QMIX pid=108284)     return WorkerSet(
(QMIX pid=108284)   File "/home/xjp/Miniconda3/envs/active-pose/lib/python3.8/site-packages/ray/rllib/evaluation/worker_set.py", line 110, in __init__
(QMIX pid=108284)     self._local_worker = self._make_worker(
(QMIX pid=108284)   File "/home/xjp/Miniconda3/envs/active-pose/lib/python3.8/site-packages/ray/rllib/evaluation/worker_set.py", line 449, in _make_worker
(QMIX pid=108284)     worker = cls(
(QMIX pid=108284)   File "/home/xjp/Miniconda3/envs/active-pose/lib/python3.8/site-packages/ray/rllib/evaluation/rollout_worker.py", line 459, in __init__
(QMIX pid=108284)     self.env = env_creator(copy.deepcopy(self.env_context))
(QMIX pid=108284)   File "train.py", line 107, in <lambda>
(QMIX pid=108284)     lambda config: make_env(config).with_agent_groups(
(QMIX pid=108284)   File "train.py", line 80, in make_env
(QMIX pid=108284)     env = make_env_impl(env_config)
(QMIX pid=108284)   File "/data/xjp/Projects/Active-Pose/activepose/control/envs.py", line 34, in make_train_env
(QMIX pid=108284)     env = make_env(**env_config)  # update config in make_env
(QMIX pid=108284)   File "/data/xjp/Projects/Active-Pose/activepose/envs/__init__.py", line 31, in make_env
(QMIX pid=108284)     env = MultiviewPose(config)
(QMIX pid=108284)   File "/data/xjp/Projects/Active-Pose/activepose/envs/base.py", line 55, in __init__
(QMIX pid=108284)     pose_estimator = initialize_streaming_pose_estimator(config)
(QMIX pid=108284)   File "/data/xjp/Projects/Active-Pose/activepose/pose/utils2d/pose/core.py", line 310, in initialize_streaming_pose_estimator
(QMIX pid=108284)     model_dict = load_pose2d_model(config, device=device)
(QMIX pid=108284)   File "/data/xjp/Projects/Active-Pose/activepose/pose/utils2d/pose/core.py", line 29, in load_pose2d_model
(QMIX pid=108284)     state_dict = torch.load(config.TEST.MODEL_FILE)
(QMIX pid=108284)   File "/home/xjp/Miniconda3/envs/active-pose/lib/python3.8/site-packages/torch/serialization.py", line 608, in load
(QMIX pid=108284)     return _legacy_load(opened_file, map_location, pickle_module, **pickle_load_args)
(QMIX pid=108284)   File "/home/xjp/Miniconda3/envs/active-pose/lib/python3.8/site-packages/torch/serialization.py", line 787, in _legacy_load
(QMIX pid=108284)     result = unpickler.load()
(QMIX pid=108284)   File "/home/xjp/Miniconda3/envs/active-pose/lib/python3.8/site-packages/torch/serialization.py", line 743, in persistent_load
(QMIX pid=108284)     deserialized_objects[root_key] = restore_location(obj, location)
(QMIX pid=108284)   File "/home/xjp/Miniconda3/envs/active-pose/lib/python3.8/site-packages/torch/serialization.py", line 175, in default_restore_location
(QMIX pid=108284)     result = fn(storage, location)
(QMIX pid=108284)   File "/home/xjp/Miniconda3/envs/active-pose/lib/python3.8/site-packages/torch/serialization.py", line 151, in _cuda_deserialize
(QMIX pid=108284)     device = validate_cuda_device(location)
(QMIX pid=108284)   File "/home/xjp/Miniconda3/envs/active-pose/lib/python3.8/site-packages/torch/serialization.py", line 135, in validate_cuda_device
(QMIX pid=108284)     raise RuntimeError('Attempting to deserialize object on a CUDA '
(QMIX pid=108284) RuntimeError: Attempting to deserialize object on a CUDA device but torch.cuda.is_available() is False. If you are running on a CPU-only machine, please use torch.load with map_location=torch.device('cpu') to map your storages to the CPU.
```

Since QMIX required agents to be grouped together, I made the following change to the environment:
```
tmp_env = make_env(tmp_env_config)

agent_group = {'camera': tmp_env.agent_ids}
obs_group = Tuple([tmp_env.observation_space for _ in agent_group["camera"]])
act_group = Tuple([tmp_env.action_space for _ in agent_group["camera"]])
tune.register_env(
    "active-pose-parallel-grouped",
    lambda config: make_env(config).with_agent_groups(
        groups=agent_group, obs_space=obs_group, act_space=act_group))
```

Here are a few things I have tried so far:

- I could run PPO (using the same make_env() callable) without running into any errors, DQN seems to work fine too.
- I can run QMIX with `local_mode = True` but the opposite cant
- While debugging QMIX with remote workers, I added a print statement inside the `make_env` function to check if `torch.cuda.is_available() ` while the remote working is initiating a new environment, but it is shown to be `False`. But for PPO, `torch.cuda.is_available() ` is shown to be `True` in remote environment creation.

The only difference is that the QMIX requires using the `with_agent_groups` wrapper, so I'm guessing this could be the source of the error?

Any help is deeply appreciated.

### Versions / Dependencies

Ray == 1.9.0, Python == 3.8.12, Pytorch == 1.10.0

### Reproduction script

Unfortunately, it is difficult for me to provide a script as our project is currently closed-source, but I am willing to check anything you would suggest.

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

**Comment by sven1977** [comment]
Hey @mickelliu , thanks for raising this issue and very cool project you are working on :)

So the error above seems to be generated when the remote workers try to create the env object(s). The stuff with PPO and DQN working, but QMIX not is very strange and I have no explanation for that. A few questions:
- In your environment you are using a pre-trained model?
- If yes, does this model need to run on a GPU?
- Do you have num_gpus_per_worker set to >0 (to make sure remote workers also have access to GPU)?

**Comment by sven1977** [comment]
Would it be possible to reproduce the issue w/o UE4, but on a mimicked random env, similar to our `rllib/examples/env/random_env::RandomEnv` examples (for which you can set the obs/action spaces)?

**Comment by mickelliu** [comment]
> 

Thanks @sven1977 for your kind responses. We are doing some cool stuff with UE4 and combined Ray Rllib into the mix. I still need to run a few tests (with your suggestions in mind) before I drew more clues out of this. But first to answer your questions:

1. Our environment is indeed using a pre-trained model, this is used to process observations received from interacting with the UE4 binary. So the Rllib agent should receive pre-processed observations from the CV model (and we did not make changes  to the pre-processing logic written in Rllib, done purely on the env level)
2. The pre-trained CV model needs a GPU for faster inference. It is a fairly deep model and I can't imagine without using any gpu.
3. I did. For PPO I set the `num_gpus_per_worker = 0.5` and `num_envs_per_worker = 4` so that each GPU hosts 2 workers, 8 vectorized envs and it works seamlessly. For QMIX I tried to either (1) the same setting as PPO that I just mentioned (2) num_gpu = 1 or num_envs_per_work = 1, but none of these worked. 

Then I figured QMIX and PPO have completed different-looking execution plan, and lots of lines of codes in QMIX's execution plan involves device changes on tensors. I am thinking could this potentially be the problem?

Thanks, 
Mickel

**Comment by sven1977** [comment]
Hey @mickelliu , thanks for your responses. This was really useful information. Yes, maybe QMIX doesn't properly move the workers to the GPUs. ...

**Comment by sven1977** [comment]
Yeah, could be a QMIX-model-specific problem here. Are you using QMIX's default model or are you providing your own custom one?

**Comment by sven1977** [comment]
Ah yeah, QMIXPolicy is NOT a `TorchPolicy` child, inherits directly from `Policy`. The generated models (main and target) are "manually" transferred to `policy.device`, whatever that is:

```
        self.device = (torch.device("cuda")
                       if torch.cuda.is_available() else torch.device("cpu"))
```

Could you print out the values of `self.device` inside `rllib.agents.qmix.qmix_policy.py::__init__` and let me know, what you see? I think the fix would be to just copy/paste the `self.device(s)` defining code from torch_policy.py into the QMIX policy as well. I'm working on a PR.

**Comment by sven1977** [comment]
More questions:
- `config.create_env_on_driver` is False (should be, otherwise, you may have no GPU available b/c Ray doesn't provide one)?
- Do you create a dummy env up-front to get the obs/action spaces so you can bundle them in a pre-defined Tuple space for the `register_env()`?

**Comment by mickelliu** [comment]
> Hey @mickelliu , thanks for your responses. This was really useful information. Yes, maybe QMIX doesn't properly move the workers to the GPUs. ...

Hi @sven1977, to follow up with more details:

  >  Are you using QMIX's default model or are you providing your own custom one?
  
  I'm using the default model (run='qmix', no custom_model)
  
  > Could you print out the values of self.device inside rllib.agents.qmix.qmix_policy.py::__init__ and let me know, what you see?
  
  **In Remote Mode:** It didn't even make it that far, it ran into the CUDA serialization error while making the env, specifically when loading the torch model onto the device. And I added a print statement `print(os.environ)` inside `make_env` to check if there are any `CUDA_VISIBLE_DEVICES` before the rolloutworker loads the torch model. Turns out it is empty, `'CUDA_VISIBLE_DEVICES': ""`
  
  **In Local Mode:** where I set local=True, and Ray actors should not have access to any GPU under this mode if I remember correctly. `'CUDA_VISIBLE_DEVICES'` is also empty here, but the program runs. `torch.load()` doesn't give me an error and QMIXTrainer is initialized. `self.device` prints `cuda`.

  
  **But in PPO**, the `'CUDA_VISIBLE_DEVICES'` is non-empty, e.g. for my first worker, it shows `'CUDA_VISIBLE_DEVICES': 0` in `os.environ`.

> config.create_env_on_driver is False (should be, otherwise, you may have no GPU available b/c Ray doesn't provide one)?

I don't believe that I have ever tuned this setting, so it is set to default.

> Do you create a dummy env up-front to get the obs/action spaces so you can bundle them in a pre-defined Tuple space for the register_env()?

Yes, it is rather inconvenient to hard-core them so I ran a dummy env just to get the observation and action spaces, then I register a new environment with them (and I did change the `run` argument under `tune.run`). This is done exactly following the example provided in this repo.

So... with the above observations, I think the reason could be that the rollout worker was not assigned any `'CUDA_VISIBLE_DEVICES'` ? I haven't read through the source code of `RolloutWorker` yet.

Anyway, thanks for your patience.
